icu4c/source/i18n/messageformat2_errors.h

```diff
@@ -0,0 +1,155 @@
+    template<>
+    Error<DynamicErrorType>::~Error();
+
+    class StaticErrors : public UObject {
+    private:
+        friend class DynamicErrors;
+
+        LocalPointer<UVector> syntaxAndDataModelErrors;
+        bool dataModelError = false;
+        bool missingSelectorAnnotationError = false;
+        bool syntaxError = false;
+
+    public:
+        StaticErrors(UErrorCode&);
+
+        void setMissingSelectorAnnotation(UErrorCode&);
+        void setDuplicateOptionName(UErrorCode&);
+        void addSyntaxError(UErrorCode&);
+        bool hasDataModelError() const { return dataModelError; }
+        bool hasSyntaxError() const { return syntaxError; }
+        bool hasMissingSelectorAnnotationError() const { return missingSelectorAnnotationError; }
+        void addError(StaticError&&, UErrorCode&);
+        void checkErrors(UErrorCode&);
+
+        const StaticError& first() const;
+        StaticErrors(const StaticErrors&, UErrorCode&);
+        StaticErrors(StaticErrors&&) noexcept;
+        virtual ~StaticErrors();
+    }; // class StaticErrors
+
+    class DynamicErrors : public UObject {
+    private:
+        const StaticErrors& staticErrors;
+        LocalPointer<UVector> resolutionAndFormattingErrors;
+        bool formattingError = false;
+        bool selectorError = false;
+        bool unknownFunctionError = false;
+        bool unresolvedVariableError = false;
+
+    public:
+        DynamicErrors(const StaticErrors&, UErrorCode&);
+
+        int32_t count() const;
+        void setSelectorError(const FunctionName&, UErrorCode&);
+        void setReservedError(UErrorCode&);
+        void setUnresolvedVariable(const VariableName&, UErrorCode&);
+        void setUnknownFunction(const FunctionName&, UErrorCode&);
+        void setFormattingError(const FunctionName&, UErrorCode&);
+        // Used when the name of the offending formatter is unknown
+        void setFormattingError(UErrorCode&);
+        void setOperandMismatchError(const FunctionName&, UErrorCode&);
+        bool hasDataModelError() const { return staticErrors.hasDataModelError(); }
+        bool hasFormattingError() const { return formattingError; }
+        bool hasSelectorError() const { return selectorError; }
+        bool hasSyntaxError() const { return staticErrors.hasSyntaxError(); }
+        bool hasUnknownFunctionError() const { return unknownFunctionError; }
+        bool hasMissingSelectorAnnotationError() const { return staticErrors.hasMissingSelectorAnnotationError(); }
+        bool hasUnresolvedVariableError() const { return unresolvedVariableError; }
+        void addError(DynamicError&&, UErrorCode&);
+        void checkErrors(UErrorCode&) const;
+        bool hasError() const;
+        bool hasStaticError() const;
+
+        const DynamicError& first() const;
+        virtual ~DynamicErrors();
+    }; // class DynamicErrors
+
+} // namespace message2
+
+U_NAMESPACE_END
+
+#endif /* #if !UCONFIG_NO_FORMATTING */
```

Comment:
Good point. I got there by copy/pasting. I just looked at some existing public header files, and they are similarly inconsistent (for example, I looked at numberformatter.h, listformatter.h, formattednumber.h). Since existing code doesn't consistently use one or the other, I would rather keep it as it is, since I don't know which way is correct.